sysvabi64/sysvabi64.rst

```diff
@@ -916,10 +923,639 @@ These follow the AAELF64_ base definition from Arm.
+
+ * IFUNC resolvers that may be called via lazy-binding must be
+   asynchronous-signal safe and thread-safe as the resolver may be
+   called from a signal handler or multi-threaded process.
+
+The order of dynamic relocation resolution across an executable and
+all its shared libraries is dependent on the implementation details of
+the dynamic linker. To avoid relocation ordering problems causing
+crashes in cross platform software the following restrictions are
+recommended:
+
+ * IFUNC resolvers must not call a function that may require IFUNC
+   initialization. If the IFUNC initialization for the called function
+   has not occured then undefined behavior results.
+
+ * IFUNC resolvers must not call a function that requires a PLT
+   entry. If the address of the IFUNC is taken the dynamic relocation
+   will be in the ``.got`` so the resolver may be called prior to
+   relocation of the ``.got.plt`` entries. In a shared library or
+   position independent executable, transferring to control to the
+   address stored in an unrelocated ``.got.plt`` entry will result in
+   undefined behavior.
+
+ * IFUNC resolvers for a given function must be defined in the same
+   translation unit as the implementations of the function.
+
+IFUNC requirements for static linkers
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Relocations to pre-emptable symbols of type ``STT_GNU_IFUNC`` are
+handled in the same way as symbols of type ``STT_FUNC``.
+
+Non pre-empatable symbols of type ``STT_GNU_FUNC`` require the
+creation of PLT entry that loads the value of:
+
+ * A ``.got`` entry for a non-branch relocation such as
+   ``R_AARCH64_ABS64``. The entry has an ``R_AARCH64_IRELATIVE``
+   dynamic relocation in ``.rela.dyn``.
+
+ * Either a ``.got`` or ``.got.plt`` entry or a for a branch
+   relocation such as ``R_AARCH64_CALL``. A ``R_AARCH64_IRELATIVE``
+   dynamic relocation is added to ``.rela.dyn`` or ``rela.plt``
+   respectively.
+
+Like ``R_AARCH64_RELATIVE`` the ``R_AARCH64_IRELATIVE`` relocation
+does not require a symbol.
+
+The static linker sets the initial contents of a ``.got`` or
+``.got.plt`` entry for a non pre-empatable IFUNC resolver to the
+address of the IFUNC resolver function.
+
+To make address equivalence of functions with IFUNC resolvers work, if
+the address of a non-preemptable ``STT_GNU_FUNC`` symbol is taken in a
+non-position independent executable. The static linker must use the
+address of the PLT for the address of the symbol. If the symbol is
+exported to the dynamic symbol table the ``st_value`` of the symbol
+must be set to the address of the PLT entry and the dynamic symbol
+table type must be set to ``STT_FUNC``.
+
+The ``.rela.dyn`` and ``rela.plt`` relocation entries for IFUNCs must
+be sorted after all non-IFUNCs. This means that for a given executable
+or shared-library the following ordering can be relied on:
+
+ * All non IFUNC ``.rela.dyn`` relocation entries will be resolved
```

Comment:
The order matches the glibc behavior, but FreeBSD rtld does not use this order. I think we probably should not make FreeBSD's implementation non-conforming.

See https://maskray.me/blog/2021-01-18-gnu-indirect-function#relocation-resolving-order for some notes about FreeBSD rtld's behavior.

Reply:
I've tried to word it as recommendations for maximum portability, rather than a requirement. For example a program written against FreeBSD `rtld` may not run if loaded by `ld.so`. That doesn't make `rtld` non compliant though.

Reply:
LGTM